cpp/ql/lib/semmle/code/cpp/security/InvalidPointerDereference/AllocationToInvalidPointer.qll

```diff
@@ -124,45 +127,83 @@ module SizeBarrier {
     )
   }
 
+  /**
+   * Holds if `left < nRight + k` holds if `g` evaluates to `edge`.
+   */
   private predicate operandGuardChecks(
-    IRGuardCondition g, Operand left, Operand right, int state, boolean edge
+    IRGuardCondition g, Operand left, DataFlow::Node right, int k, boolean edge
   ) {
-    exists(DataFlow::Node nLeft, DataFlow::Node nRight, int k |
-      nRight.asOperand() = right and
-      nLeft.asOperand() = left and
-      SizeBarrierConfig::isSink(nLeft, nRight, g, k, edge) and
-      state = getAFlowStateForNode(nRight) and
-      k <= state
+    flowTo(right) and
+    SizeBarrierConfig::isSink(DataFlow::operandNode(left), right, g, k, edge)
+  }
+
+  /**
+   * Gets an instruction `instr` that is guarded by a check such as `instr <= left + delta` where
+   * `left <= _ + k` and `left` is the "small side" of of a relational comparison that checks
+   * whether `left <= size` where `size` is the size of an allocation.
+   */
+  Instruction getABarrierInstruction0(int delta, int k) {
+    exists(IRGuardCondition g, ValueNumber value, Operand left, boolean edge, DataFlow::Node right |
+      left = value.getAUse() and
+      // We know:
+      // 1. result <= value + delta (by `bounded`)
+      // 2. value < right + k + 1 (by `operandGuardChecks`).
```

Comment:
Could we simplify some of the math here by getting rid of the `+ 1` and using `<` and `>` instead of `<=` and `>=`?

Reply:
The line right below does get rid of the `+ 1` by rewriting that inequality as `value <= right + k`. Is that not the simplification you're referring to?

Reply:
Not quite what I meant. I was wondering if we could get rid of the `+1` throughout the comment.

Reply:
Ah, I see. So you want to completely remove that intermediate step? That's fine with me. Something like in dcc1429ed8bd826f2a342d0e5a4cd2b7ca7ce6a6?

Reply:
Yeah, and probably do the same thing in the code?

Reply:
Hm, I'll see if I can. The problem is that the guards library works on inequalities of the form `x < y + k` while the range analysis library works on inequalities of the form `x <= y + k`. So the translation form `<` to `<=` needs to happen _somewhere_ 😭.

Reply:
My expectation would be that this can be fully handled by picking the correct `<`, `<=`, `>`, and `>=`, but I might be wrong here.

Reply:
1455e8bd4cb688012ae9c0d616badbe82fb3255a converts all the calculations to use `<=`. I'd like to remove the `+ 1` from `getAFlowStateForNode` because I can't quite work out why that one is needed. I'll keep thinking about this!

Reply:
Yeah, you now basically pushed down the problem, which I do not like at all.

Reply:
That's fair. Let me take another look at this once I'm back from lunch 🥘.

Reply:
To summarize: it looks to me like we should easily be able to get rid of the `+1` by using the correct comparison operators. If this is not true, I'd like to understand why that is.